Add dark mode

matt852 opened the pull request:
<!--
    Thank you for your interest in contributing to Nautobot! Please note
    that our contribution policy recommends that a feature request or bug
    report be opened for approval prior to filing a pull request. This
    helps avoid wasting time and effort on something that we might not
    be able to accept.

    Please indicate the relevant feature request or bug report below.
-->
# Closes: #729 
# What's Changed
<!--
    Please include:
    - A summary of the proposed changes
    - A sectioned breakdown for larger features under ## subheadings
    - Screenshots, example payloads where relevant:
      - Before/After for bugfixes
      - Using a new feature
-->
I added dark mode to Nautobot.  This makes minimal changes to the existing CSS configurations and continues using Bootstrap 3.

I added a small toggle feature using Javascript to the footer that allows toggling between light and dark mode.

The setting is saved client side on the client computer, and will remain set unless the user clears the browser session data (or uses something like incognito mode).
This does not make and database changes or store any theme-related data server-side.

<img width="961" alt="Screen Shot 2022-03-25 at 4 20 59 PM" src="https://user-images.githubusercontent.com/536467/160195794-c2a4412e-f2f3-426f-b4a8-260cb07dfc98.png">

<img width="1912" alt="Screen Shot 2022-03-25 at 4 20 22 PM" src="https://user-images.githubusercontent.com/536467/160195837-f8ffa259-598c-4cf5-97ed-166fa59b5fc0.png">

<img width="1918" alt="Screen Shot 2022-03-25 at 4 20 49 PM" src="https://user-images.githubusercontent.com/536467/160195858-da3a15c0-bf7c-49a2-8628-46a83ea0c339.png">

bryanculver reviewed (changes requested):
This looks like a great start @matt852!

I'd like to see this go just a bit further, referencing a guide like this: https://css-tricks.com/a-complete-guide-to-dark-mode-on-the-web/

- Using browser preference settings (I'm a day-time light mode, night-time dark mode user)

- Handling the buttons which look weird IMHO with black text.

matt852 commented:
@bryanculver I've updated the buttons, the status backgrounds, etc to always use white text.
 I updated the JS to detect the current system theme, and set Nautobot's theme to match it.  It will also change if the system theme changes.
I left in the ability for the user to select a theme from the footer button.  This enables the user to force the theme for Nautobot, and will no longer update based on the system theme (unless they use a new browser, incognito mode, etc).

glennmatthews commented on nautobot/project-static/css/base.css:
```diff
@@ -1,517 +1,544 @@
 /* Layout */
 * {
-    margin: 0;
+  margin: 0;
```
Can you please revert the indentation/whitespace changes? Makes for a lot of noise in the diffs.

matt852 replied:
My mistake, that's the auto-formatter at work.  I'll get this reverted.

glennmatthews replied:
Would still like to see this reverted. :-)

matt852 replied:
My bad, I thought I did this already.

matt852 replied:
Fixed

glennmatthews reviewed (commented):
Pretty cool! A few thoughts:

- Add an option to revert to system default theme after setting a manual theme? https://kilianvalkhof.com/2020/design/your-dark-mode-toggle-is-broken/
- Do rack-elevation SVGs render correctly in dark mode?
- How well do the Swagger (/api/docs/) and GraphiQL (/graphql/) pages render in dark mode?

matt852 commented:
I added in support for 3 options: light mode, dark mode, and system settings.  With adding a third, the toggle idea with the "light/dark" text in the footer wasn't as viable, so I created a modal that pops up.  Upon selecting a theme, the page refreshes and the modal closes, or you can simply cancel out of it.
![Screen Shot 2022-04-20 at 8 52 19 PM](https://user-images.githubusercontent.com/536467/164350306-6bf92fa4-9146-4990-b1e1-aaa12558cc8d.png)
![Screen Shot 2022-04-20 at 8 56 00 PM](https://user-images.githubusercontent.com/536467/164350425-6a25bdd9-dce8-4dfe-b6f8-c0d803da0e5a.png)

matt852 commented:
![Screen Shot 2022-04-20 at 9 16 29 PM](https://user-images.githubusercontent.com/536467/164352353-9bbf085e-ee96-4e67-95f3-848bc823d41a.png)
![Screen Shot 2022-04-20 at 9 15 42 PM](https://user-images.githubusercontent.com/536467/164352277-1973fbac-033a-4955-a882-861df19e2831.png)

matt852 commented:
![Screen Shot 2022-04-20 at 9 22 05 PM](https://user-images.githubusercontent.com/536467/164352861-2a882fbd-8f59-49fd-9b64-b1ca52d53d39.png)
![Screen Shot 2022-04-20 at 9 21 54 PM](https://user-images.githubusercontent.com/536467/164352864-8ed70792-6a4e-462c-b19e-e15b08ff986b.png)

matt852 commented:
![Screen Shot 2022-04-20 at 9 25 50 PM](https://user-images.githubusercontent.com/536467/164353239-85be210d-9fc7-4e43-91d3-60e466ddf6b5.png)
![Screen Shot 2022-04-20 at 9 25 39 PM](https://user-images.githubusercontent.com/536467/164353240-6add164a-c3a7-4905-aff4-b1fb9e96ccfe.png)

matt852 commented:
@glennmatthews @bryanculver I've addressed all of your feedback, and included screenshots for the various areas of concern.
I believe the Rack Elevations, API docs, and GraphiQL pages are all working properly now (they weren't previously...good catch!)

From my earlier comment: I added in support for 3 options: light mode, dark mode, and system settings. With adding a third, the toggle idea with the "light/dark" text in the footer wasn't as viable, so I created a modal that pops up. Upon selecting a theme, the page refreshes and the modal closes, or you can simply cancel out of it.

"System" theme is set by default right now.  I can change this if desired.

Please let me know your thoughts or concerns.  I know it's slightly different without using a fully separate light/dark stylesheet, but I believe it will work well until there's a broader UI overhaul and push towards Bootstrap 5 (or some other update).

glennmatthews commented on nautobot/core/templates/base.html:
```diff
@@ -85,22 +88,25 @@ <h4><i class="mdi mdi-alert"></i> Maintenance Mode</h4>
                             <a href="{% static 'docs/index.html' %}">Docs</a>
                         {% endif %}
                         &middot;
-                        <i class="mdi mdi-cloud-braces text-primary"></i> <a href="{% url 'api_docs' %}">API</a> &middot;
-                        <i class="mdi mdi-graphql text-primary"></i> <a href="{% url 'graphql' %}">GraphQL</a> &middot;
-                        <i class="mdi mdi-xml text-primary"></i> <a href="{{ settings.BRANDING_URLS.code }}">Code</a> &middot;
-                        <i class="mdi mdi-lifebuoy text-primary"></i> <a href="{{ settings.BRANDING_URLS.help }}">Help</a>
+                        <span class="mdi mdi-cloud-braces text-primary"></span> <a href="{% url 'api_docs' %}">API</a> &middot;
+                        <span class="mdi mdi-graphql text-primary"></span> <a href="{% url 'graphql' %}">GraphQL</a> &middot;
+                        <span class="mdi mdi-xml text-primary"></span> <a href="{{ settings.BRANDING_URLS.code }}">Code</a> &middot;
+                        <span class="mdi mdi-lifebuoy text-primary"></span> <a href="{{ settings.BRANDING_URLS.help }}">Help</a>
```
Is this a needed change or can we revert it?

matt852 replied:
It can be reverted after our conversation yesterday.

glennmatthews commented on nautobot/core/templates/base.html:
```diff
@@ -85,22 +88,25 @@ <h4><i class="mdi mdi-alert"></i> Maintenance Mode</h4>
         </div>
     </footer>
+{% include 'modals/modal_theme.html' with name='theme'%}
+<script src="{% static 'js/theme.js' %}"></script>
```
Do we need similar additions to `nautobot/core/templates/admin/base.html` so that the admin pages also support dark mode?

matt852 replied:
Yes we do.  I'll add it.
For future considerations, there's a few different "base.html" type files I've had to go in and modify.  I think ideally in the future, we consolidate these into one "true" base file.

matt852 replied:
Actually in looking at this now, it already appears the footers are out of sync.  I'll update this as well.

glennmatthews commented on nautobot/core/tests/integration/test_theme.py:
```diff
@@ -0,0 +1,63 @@
+from nautobot.utilities.testing.integration import SeleniumTestCase
+
+
+class ThemeTestCase(SeleniumTestCase):
```
This is great stuff - thank you for adding it!

glennmatthews commented on nautobot/core/urls.py:
```diff
@@ -16,6 +16,7 @@
     # Base views
     path("", HomeView.as_view(), name="home"),
     path("search/", SearchView.as_view(), name="search"),
+    path("modals/", ThemeView.as_view(), name="modals"),
```
I don't quite understand where this is used - can you explain? Isn't the modal already present in the base view, so what's the case where we need to load it on the fly?

matt852 replied:
This is a good catch.  I believe this was left over from a previous way I attempted to load the themes, but is no longer needed.  I'll remove it.

matt852 commented:
I've confirmed the Rack elevation is dark, but when selecting "Save SVG", the SVG is rendered in light mode (as requested).

bryanculver commented:
Retargeted at `next` as this is a new feature not addressing a bug.
